arangod/Cluster/HeartbeatThread.cpp

```diff
@@ -518,6 +561,195 @@ void HeartbeatThread::runDBServer() {
+  }
+
+  if (!result.successful()) {
+    if (!application_features::ApplicationServer::isStopping()) {
+      LOG_TOPIC("539fc", WARN, Logger::HEARTBEAT)
+          << "Heartbeat: Could not read from agency! status code: " << result._statusCode
+          << ", incriminating body: " << result.bodyRef() << ", timeout: " << timeout;
+    }
+  } else {
+    VPackSlice agentPool = result.slice()[0].get(".agency");
+    updateAgentPool(agentPool);
+
+    VPackSlice shutdownSlice = result.slice()[0].get(
+        std::vector<std::string>({AgencyCommManager::path(), "Shutdown"}));
+
+    if (shutdownSlice.isBool() && shutdownSlice.getBool()) {
+      ApplicationServer::server->beginShutdown();
+    }
+
+    // mop: order is actually important here...FoxxmasterQueueupdate will
+    // be set only when somebody registers some new queue stuff (for example
+    // on a different coordinator than this one)... However when we are just
+    // about to become the new foxxmaster we must immediately refresh our
+    // queues this is done in ServerState...if queueupdate is set after
+    // foxxmaster the change will be reset again
+    VPackSlice foxxmasterQueueupdateSlice = result.slice()[0].get(std::vector<std::string>(
+        {AgencyCommManager::path(), "Current", "FoxxmasterQueueupdate"}));
+
+    if (foxxmasterQueueupdateSlice.isBool()) {
+      ServerState::instance()->setFoxxmasterQueueupdate(
+          foxxmasterQueueupdateSlice.getBool());
+    }
+
+    VPackSlice foxxmasterSlice = result.slice()[0].get(std::vector<std::string>(
+        {AgencyCommManager::path(), "Current", "Foxxmaster"}));
+
+    if (foxxmasterSlice.isString() && foxxmasterSlice.getStringLength() != 0) {
+      ServerState::instance()->setFoxxmaster(foxxmasterSlice.copyString());
+    } else {
+      auto state = ServerState::instance();
+      VPackBuilder myIdBuilder;
+      myIdBuilder.add(VPackValue(state->getId()));
+
+      auto updateMaster = agency.casValue("/Current/Foxxmaster", foxxmasterSlice,
+                                           myIdBuilder.slice(), 0, 10.0);
+      if (updateMaster.successful()) {
+        // We won the race we are the master
+        ServerState::instance()->setFoxxmaster(state->getId());
+      }
+      agency.increment("Current/Version");
+    }
+
+    VPackSlice versionSlice = result.slice()[0].get(std::vector<std::string>(
+        {AgencyCommManager::path(), "Plan", "Version"}));
+
+    if (versionSlice.isInteger()) {
+      // there is a plan version
+
+      uint64_t planVersion = 0;
+      try {
+        planVersion = versionSlice.getUInt();
+      } catch (...) {
+      }
+
+      if (planVersion > _lastPlanVersionNoticed) {
+        LOG_TOPIC("7c80a", TRACE, Logger::HEARTBEAT)
+            << "Found planVersion " << planVersion << " which is newer than "
+            << _lastPlanVersionNoticed;
+        if (handlePlanChangeCoordinator(planVersion)) {
+          _lastPlanVersionNoticed = planVersion;
```

Comment:
I don't know this part of the code very well, so this may be a stupid remark, but this unconditional store to `_lastPlanVersionNoticed` looks suspicious to me.
The comment at the begin of the function states that this function is executed by a scheduler thread and we have to "protect ourselves against concurrent accesses", so I assume that this function can be called multiple times concurrently. But then it is not guaranteed that at the time of this assignment, `_lastPlanVersionNoticed` still holds the same value we previously saw in the comparison in line 659. E.g., another thread could have updated `_lastPlanVersionNoticed` with an even greater plan version which we would now override.
Am I missing something?

Reply:
Hi Manuel, thanks for looking.
Actually, the logic in the Heartbeat thread implies that this method is only ever called once at a time. A new run is triggered in the heartbeat if an atomic variable has the value 0. The heartbeat then sets this to 1 and schedules a closure on the Scheduler, which calls this method and, as last action, resets the atomic to 0.
Therefore these members are a safe place to keep the information about the last processed Plan version and Current version.
In the not too distant future we want to move to an incremental update of Plan and Current anyway and then we will delete this code in its entirety. Therefore, I wanted to keep the changes minimal and just moved the local variables to members. I will add a comment about this.

Reply:
Thanks for the clarification!